File: apps/web/src/features/block-canvas/store/canvas-loro.ts
import { LoroDoc } from 'loro-crdt';

const STORAGE_KEY = 'conation:canvas-loro';
const MAX_DOCS = 40;
const MAX_UPDATES = 64;

export type CanvasLoroJson = {
  nodes?: unknown[];
  edges?: unknown[];
  groups?: unknown[];
};

type CanvasLoroRecord = {
  documentId: string;
  updates: string[];
  savedAt: number;
};

function bytesToB64(bytes: Uint8Array): string {
  let binary = '';
  for (const b of bytes) binary += String.fromCharCode(b);
  return btoa(binary);
}

function b64ToBytes(value: string): Uint8Array {
  const binary = atob(value);
  const out = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) out[i] = binary.charCodeAt(i);
  return out;
}

function entityId(item: unknown): string | undefined {
  if (!item || typeof item !== 'object' || Array.isArray(item)) return;
  const id = (item as { id?: unknown }).id;
  return typeof id === 'string' ? id : undefined;
}

export type CanvasOp =
  | { op: 'upsertNode'; node: Record<string, unknown> }
  | { op: 'deleteNode'; id: string }
  | { op: 'moveNode'; id: string; x: number; y: number }
  | { op: 'updateNode'; id: string; patch: Record<string, unknown> }
  | { op: 'upsertEdge'; edge: Record<string, unknown> }
  | { op: 'deleteEdge'; id: string };

function upsertEntity(
  items: unknown[],
  item: Record<string, unknown>
): unknown[] {
  const id = entityId(item);
  if (!id) return items;
  const next = items.filter((entry) => entityId(entry) !== id);
  next.push(item);
  return next;
}

function deleteEntity(items: unknown[], id: string): unknown[] {
  return items.filter((entry) => entityId(entry) !== id);
}

function patchNode(
  items: unknown[],
  id: string,
  patch: Record<string, unknown>
): unknown[] {
  return items.map((entry) => {
    if (entityId(entry) !== id || !entry || typeof entry !== 'object') {
      return entry;
    }
    const { id: _ignored, ...rest } = patch;
    return { ...(entry as Record<string, unknown>), ...rest, id };
  });
}

function itemsById(
  items: unknown[] | undefined
): Map<string, Record<string, unknown>> {
  const map = new Map<string, Record<string, unknown>>();
  for (const item of items ?? []) {
    const id = entityId(item);
    if (!id || !item || typeof item !== 'object' || Array.isArray(item)) {
      continue;
    }
    map.set(id, item as Record<string, unknown>);
  }
  return map;
}

function sameJson(left: unknown, right: unknown): boolean {
  return JSON.stringify(left) === JSON.stringify(right);
}

function readMapEntity(
  map: ReturnType<LoroDoc['getMap']>,
  id: string
): Record<string, unknown> | undefined {
  const raw = (map.toJSON() as Record<string, unknown>)?.[id];
  if (typeof raw !== 'string') return;
  try {
    const parsed = JSON.parse(raw) as unknown;
    if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) {
      return;
    }
    return parsed as Record<string, unknown>;
  } catch {
    return;
  }
}

/** Diff two boards into editor/agent node-edge ops (untouched ids omitted). */
export function opsFromBoardDiff(
  previous: CanvasLoroJson,
  next: CanvasLoroJson
): CanvasOp[] {
  const ops: CanvasOp[] = [];
  const prevNodes = itemsById(previous.nodes);
  const nextNodes = itemsById(next.nodes);
  for (const [id, node] of nextNodes) {
    const prev = prevNodes.get(id);
    if (!prev) {
      ops.push({ op: 'upsertNode', node });
      continue;
    }
    if (sameJson(prev, node)) continue;
    ops.push({ op: 'upsertNode', node });
  }
  for (const id of prevNodes.keys()) {
    if (!nextNodes.has(id)) ops.push({ op: 'deleteNode', id });
  }

  const prevEdges = itemsById(previous.edges);
  const nextEdges = itemsById(next.edges);
  for (const [id, edge] of nextEdges) {
    const prev = prevEdges.get(id);
    if (!prev) {
      ops.push({ op: 'upsertEdge', edge });
      continue;
    }
    if (sameJson(prev, edge)) continue;
    ops.push({ op: 'upsertEdge', edge });
  }
  for (const id of prevEdges.keys()) {
    if (!nextEdges.has(id)) ops.push({ op: 'deleteEdge', id });
  }
  return ops;
}

/** Apply editor-style node/edge ops onto a board (agent + tests). */
export function applyCanvasOps(
  json: CanvasLoroJson,
  ops: CanvasOp[]
): CanvasLoroJson {
  let nodes = [...(json.nodes ?? [])];
  let edges = [...(json.edges ?? [])];
  for (const op of ops) {
    switch (op.op) {
      case 'upsertNode':
        nodes = upsertEntity(nodes, op.node);
        break;
      case 'deleteNode':
        nodes = deleteEntity(nodes, op.id);
        break;
      case 'moveNode':
        nodes = patchNode(nodes, op.id, { x: op.x, y: op.y });
        break;
      case 'updateNode':
        nodes = patchNode(nodes, op.id, op.patch);
        break;
      case 'upsertEdge':
        edges = upsertEntity(edges, op.edge);
        break;
      case 'deleteEdge':
        edges = deleteEntity(edges, op.id);
        break;
      default: {
        const _never: never = op;
        void _never;
      }
    }
  }
  return { ...json, nodes, edges };
}

/** Mutate only named Loro entities (same as agent `canvasOps` apply-update). */
export function applyCanvasOpsToDoc(doc: LoroDoc, ops: CanvasOp[]): void {
  const nodes = doc.getMap('nodes');
  const edges = doc.getMap('edges');
  for (const op of ops) {
    switch (op.op) {
      case 'upsertNode': {
        const id = entityId(op.node);
        if (id) nodes.set(id, JSON.stringify(op.node));
        break;
      }
      case 'deleteNode':
        nodes.delete(op.id);
        break;
      case 'moveNode': {
        const node = readMapEntity(nodes, op.id);
        if (node) {
          nodes.set(
            op.id,
            JSON.stringify({ ...node, x: op.x, y: op.y, id: op.id })
          );
        }
        break;
      }
      case 'updateNode': {
        const node = readMapEntity(nodes, op.id);
        if (node) {
          const { id: _ignored, ...rest } = op.patch;
          nodes.set(op.id, JSON.stringify({ ...node, ...rest, id: op.id }));
        }
        break;
      }
      case 'upsertEdge': {
        const id = entityId(op.edge);
        if (id) edges.set(id, JSON.stringify(op.edge));
        break;
      }
      case 'deleteEdge':
        edges.delete(op.id);
        break;
      default: {
        const _never: never = op;
        void _never;
      }
    }
  }
}

function applyDirtyGroups(
  doc: LoroDoc,
  previous: unknown[] | undefined,
  next: unknown[] | undefined
): void {
  const map = doc.getMap('groups');
  const prev = itemsById(previous);
  const nxt = itemsById(next);
  for (const [id, item] of nxt) {
    const prior = prev.get(id);
    if (prior && sameJson(prior, item)) continue;
    map.set(id, JSON.stringify(item));
  }
  for (const id of prev.keys()) {
    if (!nxt.has(id)) map.delete(id);
  }
}

function syncMap(map: ReturnType<LoroDoc['getMap']>, items: unknown[]): void {
  const keep = new Set<string>();
  for (const item of items) {
    const id = entityId(item);
    if (!id) continue;
    keep.add(id);
    map.set(id, JSON.stringify(item));
  }
  const existing = Object.keys((map.toJSON() as Record<string, unknown>) ?? {});
  for (const id of existing) {
    if (!keep.has(id)) map.delete(id);
  }
}

function applyBoard(doc: LoroDoc, json: CanvasLoroJson): void {
  syncMap(doc.getMap('nodes'), json.nodes ?? []);
  syncMap(doc.getMap('edges'), json.edges ?? []);
  if (json.groups) syncMap(doc.getMap('groups'), json.groups);
  doc.commit();
}

function mapToEntities(doc: LoroDoc, name: string): unknown[] {
  const raw = doc.getMap(name).toJSON() as Record<string, unknown>;
  const items: unknown[] = [];
  for (const value of Object.values(raw ?? {})) {
    if (typeof value !== 'string') continue;
    try {
      items.push(JSON.parse(value));
    } catch {
      /* skip corrupt entry */
    }
  }
  items.sort((a, b) => {
    const left = entityId(a) ?? '';
    const right = entityId(b) ?? '';
    return left.localeCompare(right);
  });
  return items;
}

export function boardFromDoc(doc: LoroDoc): CanvasLoroJson {
  return {
    nodes: mapToEntities(doc, 'nodes'),
    edges: mapToEntities(doc, 'edges'),
    groups: mapToEntities(doc, 'groups'),
  };
}

/** Encode a board as a Loro snapshot for sync-service initialize. */
export function snapshotFromJson(json: CanvasLoroJson): Uint8Array {
  const doc = new LoroDoc();
  doc.setPeerId(1n);
  applyBoard(doc, json);
  return doc.export({ mode: 'snapshot' });
}

/** Encode one peer's board as a Loro update. */
export function encodeCanvasLoroUpdate(
  json: CanvasLoroJson,
  peerId: bigint
): Uint8Array {
  const doc = new LoroDoc();
  doc.setPeerId(peerId);
  applyBoard(doc, json);
  return doc.export({ mode: 'update' });
}

/**
 * Replay prior updates, apply dirty node/edge ops (and dirty groups), and
 * export the incremental Loro update. Whole-board `applyBoard` is the
 * fallback when there is no prior WAL to diff against.
 */
export function encodeCanvasLoroDiff(
  previousUpdates: Uint8Array[],
  json: CanvasLoroJson,
  peerId: bigint
): Uint8Array {
  const doc = new LoroDoc();
  doc.setPeerId(peerId);
  for (const update of previousUpdates) {
    doc.import(update);
  }
  const from = doc.version();
  if (previousUpdates.length === 0) {
    applyBoard(doc, json);
    return doc.export({ mode: 'update', from });
  }
  const previous = boardFromDoc(doc);
  applyCanvasOpsToDoc(doc, opsFromBoardDiff(previous, json));
  if (json.groups) applyDirtyGroups(doc, previous.groups, json.groups);
  doc.commit();
  return doc.export({ mode: 'update', from });
}

/** Import Loro updates from independent peers and export merged board JSON. */
export function mergeCanvasLoroUpdates(updates: Uint8Array[]): CanvasLoroJson {
  const doc = new LoroDoc();
  for (const update of updates) {
    doc.import(update);
  }
  return boardFromDoc(doc);
}

export function mergeCanvasBoards(
  left: CanvasLoroJson,
  right: CanvasLoroJson
): CanvasLoroJson {
  return mergeCanvasLoroUpdates([
    encodeCanvasLoroUpdate(left, 1n),
    encodeCanvasLoroUpdate(right, 2n),
  ]);
}

function readStore(): Record<string, CanvasLoroRecord> {
  if (typeof localStorage === 'undefined') return {};
  try {
    const raw = localStorage.getItem(STORAGE_KEY);
    if (!raw) return {};
    const parsed = JSON.parse(raw) as unknown;
    if (!parsed || typeof parsed !== 'object' || Array.isArray(parsed)) {
      return {};
    }
    return parsed as Record<string, CanvasLoroRecord>;
  } catch {
    return {};
  }
}

function writeStore(store: Record<string, CanvasLoroRecord>): boolean {
  if (typeof localStorage === 'undefined') return false;
  const entries = Object.values(store).sort((a, b) => b.savedAt - a.savedAt);
  const next: Record<string, CanvasLoroRecord> = {};
  for (const row of entries.slice(0, MAX_DOCS)) {
    next[row.documentId] = {
      ...row,
      updates: row.updates.slice(-MAX_UPDATES),
    };
  }
  try {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(next));
    return true;
  } catch {
    return false;
  }
}

function storedUpdates(documentId: string): Uint8Array[] {
  const row = readStore()[documentId];
  if (!row?.updates.length) return [];
  return row.updates.map(b64ToBytes);
}

/** Append a Loro update for this board so later peers can merge. */
export function recordCanvasLoro(
  documentId: string,
  json: CanvasLoroJson
): Uint8Array | null {
  if (!documentId) return null;
  const previous = storedUpdates(documentId);
  const update = encodeCanvasLoroDiff(
    previous,
    json,
    BigInt(Date.now() % Number.MAX_SAFE_INTEGER) || 1n
  );
  const store = readStore();
  const existing = store[documentId];
  store[documentId] = {
    documentId,
    updates: [...(existing?.updates ?? []), bytesToB64(update)],
    savedAt: Date.now(),
  };
  if (!writeStore(store)) return null;
  return update;
}

/** Merged board from persisted Loro updates, if any. */
export function peekCanvasLoro(documentId: string): CanvasLoroJson | null {
  const row = readStore()[documentId];
  if (!row?.updates.length) return null;
  const updates = row.updates.map(b64ToBytes);
  return mergeCanvasLoroUpdates(updates);
}

export function clearCanvasLoro(documentId: string): void {
  const store = readStore();
  if (!(documentId in store)) return;
  delete store[documentId];
  writeStore(store);
}

/** Test helper. */
export function clearAllCanvasLoro(): void {
  if (typeof localStorage === 'undefined') return;
  localStorage.removeItem(STORAGE_KEY);
}

export function hasCanvasLoro(documentId: string): boolean {
  return (readStore()[documentId]?.updates.length ?? 0) > 0;
}

File: crates/documents/src/domain/canvas_loro.rs
//! Persist canvas `{nodes, edges}` JSON as a Loro CRDT so concurrent
//! board edits merge instead of last-write-wins on the whole file.

use std::borrow::Cow;
use std::collections::HashSet;

use loro::{ExportMode, LoroDoc, LoroValue};
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};

const NODES: &str = "nodes";
const EDGES: &str = "edges";
const GROUPS: &str = "groups";

/// Empty board matching Create canvas / CreateDocument.
pub const EMPTY_BOARD_JSON: &str = r#"{"nodes":[],"edges":[]}"#;

/// Failed to encode or merge a canvas Loro document.
#[derive(Debug, thiserror::Error)]
pub enum CanvasLoroError {
    /// JSON is not a canvas board object.
    #[error("canvas content must be a JSON object with nodes and edges arrays")]
    InvalidBoard,
    /// A node-level op is missing a required id or entity object.
    #[error("{0}")]
    InvalidOp(String),
    /// Loro encode/import failed.
    #[error("{0}")]
    Loro(String),
}

/// One node/edge mutation matching the human editor store
/// (`createNode` / `updateNode` / `delete` / `createEdge`).
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[cfg_attr(feature = "ai_tools", derive(schemars::JsonSchema))]
#[serde(tag = "op", rename_all = "camelCase")]
pub enum CanvasOp {
    /// Insert or replace a node by `id` (editor `createNode` / full `updateNode`).
    UpsertNode {
        /// Node object; must include `id`.
        node: Value,
    },
    /// Remove a node by id (editor `delete`). Missing ids are a no-op.
    DeleteNode {
        /// Node id.
        id: String,
    },
    /// Set `x`/`y` on an existing node (editor move). Missing ids are a no-op.
    MoveNode {
        /// Node id.
        id: String,
        /// Board x.
        x: f64,
        /// Board y.
        y: f64,
    },
    /// Merge fields onto an existing node (editor `updateNode` patch).
    UpdateNode {
        /// Node id.
        id: String,
        /// Fields to merge; `id` in the patch is ignored.
        #[serde(default)]
        patch: Map<String, Value>,
    },
    /// Insert or replace an edge by `id` (editor `createEdge`).
    UpsertEdge {
        /// Edge object; must include `id`.
        edge: Value,
    },
    /// Remove an edge by id. Missing ids are a no-op.
    DeleteEdge {
        /// Edge id.
        id: String,
    },
}

/// Encode a canvas board as a Loro snapshot (peer 1).
pub fn snapshot_from_json(json: &str) -> Result<Vec<u8>, CanvasLoroError> {
    snapshot_from_json_with_peer(json, 1)
}

/// Encode a canvas board as a Loro snapshot from a specific peer.
pub fn snapshot_from_json_with_peer(json: &str, peer: u64) -> Result<Vec<u8>, CanvasLoroError> {
    let value: Value = serde_json::from_str(json).map_err(|_| CanvasLoroError::InvalidBoard)?;
    let doc = doc_from_board(&value, peer)?;
    doc.export(ExportMode::Snapshot)
        .map_err(|e| CanvasLoroError::Loro(e.to_string()))
}

/// Import two independently encoded boards and return merged canvas JSON.
///
/// Nodes/edges/groups are keyed by `id`, so two peers adding different
/// shapes keep both. Same-id keys last-write-win via Loro's map CRDT.
/// Removals only drop an entity when encoded as a Loro map delete
/// ([`update_from_json_onto_snapshot`]), not when a later snapshot simply
/// omits the id.
pub fn merge_json_boards(left: &str, right: &str) -> Result<String, CanvasLoroError> {
    let left_snap = snapshot_from_json_with_peer(left, 1)?;
    let right_snap = snapshot_from_json_with_peer(right, 2)?;
    json_from_snapshot(&merge_snapshots(&left_snap, &right_snap)?)
}

/// What sync-service needs so a canvas edit is visible on the live WS.
#[derive(Debug)]
pub enum CanvasSyncSeed {
    /// No Loro session yet — POST `/initialize`.
    Initialize(Vec<u8>),
    /// Session already exists — POST incremental `/apply`.
    ApplyUpdate(Vec<u8>),
}

/// Encode an initialize snapshot or an incremental apply-update.
///
/// Boards that never got `initialize_from_snapshot` take the initialize
/// path. Later agent overwrites take apply-update so live peers see the
/// same ops the editor `pushUpdate`s.
pub fn canvas_sync_seed(
    existing_snapshot: Option<&[u8]>,
    json: &str,
) -> Result<CanvasSyncSeed, CanvasLoroError> {
    match existing_snapshot {
        Some(snapshot) if !snapshot.is_empty() => Ok(CanvasSyncSeed::ApplyUpdate(
            update_from_json_onto_snapshot(snapshot, json)?,
        )),
        _ => Ok(CanvasSyncSeed::Initialize(snapshot_from_json(json)?)),
    }
}

/// Encode initialize vs apply-update for node-level ops.
///
/// When a session exists, only the named entities are upserted/deleted in
/// Loro — untouched nodes are not rewritten. Missing sessions initialize
/// from the already-applied board JSON.
pub fn canvas_sync_seed_ops(
    existing_snapshot: Option<&[u8]>,
    json: &str,
    ops: &[CanvasOp],
) -> Result<CanvasSyncSeed, CanvasLoroError> {
    match existing_snapshot {
        Some(snapshot) if !snapshot.is_empty() => Ok(CanvasSyncSeed::ApplyUpdate(
            update_from_ops_onto_snapshot(snapshot, ops)?,
        )),
        _ => Ok(CanvasSyncSeed::Initialize(snapshot_from_json(json)?)),
    }
}

/// Apply editor-style node/edge ops onto a canvas JSON board.
pub fn apply_ops_to_json(json: &str, ops: &[CanvasOp]) -> Result<String, CanvasLoroError> {
    let mut value: Value = serde_json::from_str(json).map_err(|_| CanvasLoroError::InvalidBoard)?;
    apply_ops_to_value(&mut value, ops)?;
    serde_json::to_string(&value).map_err(|e| CanvasLoroError::Loro(e.to_string()))
}

/// Apply ops onto an existing snapshot and export a Loro update.
pub fn update_from_ops_onto_snapshot(
    snapshot: &[u8],
    ops: &[CanvasOp],
) -> Result<Vec<u8>, CanvasLoroError> {
    update_from_ops_onto_snapshot_with_peer(snapshot, ops, 1)
}

/// [`update_from_ops_onto_snapshot`] with an explicit Loro peer id.
pub fn update_from_ops_onto_snapshot_with_peer(
    snapshot: &[u8],
    ops: &[CanvasOp],
    peer: u64,
) -> Result<Vec<u8>, CanvasLoroError> {
    let doc = LoroDoc::new();
    doc.set_peer_id(peer)
        .map_err(|e| CanvasLoroError::Loro(e.to_string()))?;
    doc.import(snapshot)
        .map_err(|e| CanvasLoroError::Loro(e.to_string()))?;
    let from = doc.oplog_vv();
    apply_ops_to_doc(&doc, ops)?;
    doc.commit();
    doc.export(ExportMode::Updates {
        from: Cow::Owned(from),
    })
    .map_err(|e| CanvasLoroError::Loro(e.to_string()))
}

/// Apply a later board onto an existing snapshot and export a Loro *update*
/// (the live WS payload after `initialize_from_snapshot`).
///
/// Entities present in `snapshot` but missing from `json` are deleted so
/// the tombstone merges with peers that still have the old shape.
pub fn update_from_json_onto_snapshot(
    snapshot: &[u8],
    json: &str,
) -> Result<Vec<u8>, CanvasLoroError> {
    update_from_json_onto_snapshot_with_peer(snapshot, json, 1)
}

/// [`update_from_json_onto_snapshot`] with an explicit Loro peer id.
pub fn update_from_json_onto_snapshot_with_peer(
    snapshot: &[u8],
    json: &str,
    peer: u64,
) -> Result<Vec<u8>, CanvasLoroError> {
    let value: Value = serde_json::from_str(json).map_err(|_| CanvasLoroError::InvalidBoard)?;
    let obj = value.as_object().ok_or(CanvasLoroError::InvalidBoard)?;
    if !obj.get(NODES).is_some_and(Value::is_array) || !obj.get(EDGES).is_some_and(Value::is_array)
    {
        return Err(CanvasLoroError::InvalidBoard);
    }

    let doc = LoroDoc::new();
    doc.set_peer_id(peer)
        .map_err(|e| CanvasLoroError::Loro(e.to_string()))?;
    doc.import(snapshot)
        .map_err(|e| CanvasLoroError::Loro(e.to_string()))?;
    let from = doc.oplog_vv();
    sync_entities(&doc, NODES, obj.get(NODES))?;
    sync_entities(&doc, EDGES, obj.get(EDGES))?;
    sync_entities(&doc, GROUPS, obj.get(GROUPS))?;
    doc.commit();
    doc.export(ExportMode::Updates {
        from: Cow::Owned(from),
    })
    .map_err(|e| CanvasLoroError::Loro(e.to_string()))
}

fn merge_snapshots(left: &[u8], right: &[u8]) -> Result<Vec<u8>, CanvasLoroError> {
    let doc = LoroDoc::new();
    doc.import(left)
        .map_err(|e| CanvasLoroError::Loro(e.to_string()))?;
    doc.import(right)
        .map_err(|e| CanvasLoroError::Loro(e.to_string()))?;
    doc.export(ExportMode::Snapshot)
        .map_err(|e| CanvasLoroError::Loro(e.to_string()))
}

pub(crate) fn json_from_snapshot(snapshot: &[u8]) -> Result<String, CanvasLoroError> {
    let doc = LoroDoc::new();
    doc.import(snapshot)
        .map_err(|e| CanvasLoroError::Loro(e.to_string()))?;
    let board = board_from_doc(&doc)?;
    serde_json::to_string(&board).map_err(|e| CanvasLoroError::Loro(e.to_string()))
}

fn doc_from_board(value: &Value, peer: u64) -> Result<LoroDoc, CanvasLoroError> {
    let obj = value.as_object().ok_or(CanvasLoroError::InvalidBoard)?;
    if !obj.get(NODES).is_some_and(Value::is_array) || !obj.get(EDGES).is_some_and(Value::is_array)
    {
        return Err(CanvasLoroError::InvalidBoard);
    }

    let doc = LoroDoc::new();
    doc.set_peer_id(peer)
        .map_err(|e| CanvasLoroError::Loro(e.to_string()))?;
    sync_entities(&doc, NODES, obj.get(NODES))?;
    sync_entities(&doc, EDGES, obj.get(EDGES))?;
    sync_entities(&doc, GROUPS, obj.get(GROUPS))?;
    doc.commit();
    Ok(doc)
}

fn sync_entities(
    doc: &LoroDoc,
    container: &str,
    value: Option<&Value>,
) -> Result<(), CanvasLoroError> {
    let Some(items) = value.and_then(Value::as_array) else {
        return Ok(());
    };
    let map = doc.get_map(container);
    let mut keep = HashSet::new();
    for item in items {
        let id = item
            .get("id")
            .and_then(Value::as_str)
            .ok_or(CanvasLoroError::InvalidBoard)?;
        keep.insert(id.to_string());
        map.insert(id, item.to_string())
            .map_err(|e| CanvasLoroError::Loro(e.to_string()))?;
    }
    let existing = match map.get_deep_value() {
        LoroValue::Map(entries) => entries
            .iter()
            .map(|(id, _)| id.to_string())
            .collect::<Vec<_>>(),
        _ => Vec::new(),
    };
    for id in existing {
        if !keep.contains(&id) {
            map.delete(&id)
                .map_err(|e| CanvasLoroError::Loro(e.to_string()))?;
        }
    }
    Ok(())
}

fn require_board_arrays(value: &Value) -> Result<(), CanvasLoroError> {
    let obj = value.as_object().ok_or(CanvasLoroError::InvalidBoard)?;
    if !obj.get(NODES).is_some_and(Value::is_array) || !obj.get(EDGES).is_some_and(Value::is_array)
    {
        return Err(CanvasLoroError::InvalidBoard);
    }
    Ok(())
}

fn entity_id(item: &Value) -> Result<&str, CanvasLoroError> {
    item.get("id").and_then(Value::as_str).ok_or_else(|| {
        CanvasLoroError::InvalidOp("canvas op entity must include a string id".to_string())
    })
}

fn json_number(value: f64) -> Value {
    Value::Number(serde_json::Number::from_f64(value).unwrap_or_else(|| 0.into()))
}

fn upsert_entity(items: &mut Vec<Value>, item: Value) -> Result<(), CanvasLoroError> {
    let id = entity_id(&item)?.to_string();
    if let Some(existing) = items
        .iter_mut()
        .find(|n| n.get("id").and_then(Value::as_str) == Some(id.as_str()))
    {
        *existing = item;
    } else {
        items.push(item);
    }
    Ok(())
}

fn delete_entity(items: &mut Vec<Value>, id: &str) {
    items.retain(|n| n.get("id").and_then(Value::as_str) != Some(id));
}

fn apply_ops_to_value(value: &mut Value, ops: &[CanvasOp]) -> Result<(), CanvasLoroError> {
    require_board_arrays(value)?;
    let obj = value.as_object_mut().ok_or(CanvasLoroError::InvalidBoard)?;
    for op in ops {
        match op {
            CanvasOp::UpsertNode { node } => {
                let items = obj
                    .get_mut(NODES)
                    .and_then(Value::as_array_mut)
                    .ok_or(CanvasLoroError::InvalidBoard)?;
                upsert_entity(items, node.clone())?;
            }
            CanvasOp::DeleteNode { id } => {
                let items = obj
                    .get_mut(NODES)
                    .and_then(Value::as_array_mut)
                    .ok_or(CanvasLoroError::InvalidBoard)?;
                delete_entity(items, id);
            }
            CanvasOp::MoveNode { id, x, y } => {
                let items = obj
                    .get_mut(NODES)
                    .and_then(Value::as_array_mut)
                    .ok_or(CanvasLoroError::InvalidBoard)?;
                if let Some(node) = items
                    .iter_mut()
                    .find(|n| n.get("id").and_then(Value::as_str) == Some(id.as_str()))
                    && let Some(node) = node.as_object_mut()
                {
                    node.insert("x".to_string(), json_number(*x));
                    node.insert("y".to_string(), json_number(*y));
                }
            }
            CanvasOp::UpdateNode { id, patch } => {
                let items = obj
                    .get_mut(NODES)
                    .and_then(Value::as_array_mut)
                    .ok_or(CanvasLoroError::InvalidBoard)?;
                if let Some(node) = items
                    .iter_mut()
                    .find(|n| n.get("id").and_then(Value::as_str) == Some(id.as_str()))
                    && let Some(node) = node.as_object_mut()
                {
                    for (key, val) in patch {
                        if key == "id" {
                            continue;
                        }
                        node.insert(key.clone(), val.clone());
                    }
                }
            }
            CanvasOp::UpsertEdge { edge } => {
                let items = obj
                    .get_mut(EDGES)
                    .and_then(Value::as_array_mut)
                    .ok_or(CanvasLoroError::InvalidBoard)?;
                upsert_entity(items, edge.clone())?;
            }
            CanvasOp::DeleteEdge { id } => {
                let items = obj
                    .get_mut(EDGES)
                    .and_then(Value::as_array_mut)
                    .ok_or(CanvasLoroError::InvalidBoard)?;
                delete_entity(items, id);
            }
        }
    }
    Ok(())
}

fn entity_json(doc: &LoroDoc, container: &str, id: &str) -> Option<Value> {
    let LoroValue::Map(entries) = doc.get_map(container).get_deep_value() else {
        return None;
    };
    for (key, value) in entries.iter() {
        if key.as_str() != id {
            continue;
        }
        let LoroValue::String(raw) = value else {
            return None;
        };
        return serde_json::from_str(raw).ok();
    }
    None
}

fn upsert_map(doc: &LoroDoc, container: &str, item: &Value) -> Result<(), CanvasLoroError> {
    let id = entity_id(item)?;
    doc.get_map(container)
        .insert(id, item.to_string())
        .map_err(|e| CanvasLoroError::Loro(e.to_string()))
}

fn delete_map(doc: &LoroDoc, container: &str, id: &str) -> Result<(), CanvasLoroError> {
    doc.get_map(container)
        .delete(id)
        .map_err(|e| CanvasLoroError::Loro(e.to_string()))
}

fn apply_ops_to_doc(doc: &LoroDoc, ops: &[CanvasOp]) -> Result<(), CanvasLoroError> {
    for op in ops {
        match op {
            CanvasOp::UpsertNode { node } => upsert_map(doc, NODES, node)?,
            CanvasOp::DeleteNode { id } => delete_map(doc, NODES, id)?,
            CanvasOp::MoveNode { id, x, y } => {
                if let Some(mut node) = entity_json(doc, NODES, id)
                    && let Some(obj) = node.as_object_mut()
                {
                    obj.insert("x".to_string(), json_number(*x));
                    obj.insert("y".to_string(), json_number(*y));
                    upsert_map(doc, NODES, &Value::Object(obj.clone()))?;
                }
            }
            CanvasOp::UpdateNode { id, patch } => {
                if let Some(mut node) = entity_json(doc, NODES, id)
                    && let Some(obj) = node.as_object_mut()
                {
                    for (key, val) in patch {
                        if key == "id" {
                            continue;
                        }
                        obj.insert(key.clone(), val.clone());
                    }
                    upsert_map(doc, NODES, &Value::Object(obj.clone()))?;
                }
            }
            CanvasOp::UpsertEdge { edge } => upsert_map(doc, EDGES, edge)?,
            CanvasOp::DeleteEdge { id } => delete_map(doc, EDGES, id)?,
        }
    }
    Ok(())
}

fn board_from_doc(doc: &LoroDoc) -> Result<Value, CanvasLoroError> {
    let mut obj = Map::new();
    obj.insert(NODES.to_string(), entities_from_map(doc, NODES)?);
    obj.insert(EDGES.to_string(), entities_from_map(doc, EDGES)?);
    obj.insert(GROUPS.to_string(), entities_from_map(doc, GROUPS)?);
    Ok(Value::Object(obj))
}

fn entities_from_map(doc: &LoroDoc, container: &str) -> Result<Value, CanvasLoroError> {
    let LoroValue::Map(entries) = doc.get_map(container).get_deep_value() else {
        return Ok(Value::Array(Vec::new()));
    };
    let mut items = Vec::new();
    for (_id, value) in entries.iter() {
        let LoroValue::String(raw) = value else {
            continue;
        };
        if let Ok(parsed) = serde_json::from_str::<Value>(raw) {
            items.push(parsed);
        }
    }
    items.sort_by(|a, b| {
        a.get("id")
            .and_then(Value::as_str)
            .cmp(&b.get("id").and_then(Value::as_str))
    });
    Ok(Value::Array(items))
}

#[cfg(test)]
mod test;
